extensions/avro/src/main/java/com/hazelcast/jet/avro/AvroSerializerHooks.java

```diff
@@ -64,4 +179,216 @@ public boolean isOverwritable() {
             return false;
         }
     }
+
+    private abstract static class LazyImmutableContainer<T extends GenericContainer> implements GenericContainer {
+        private final byte[] serialized;
+        private final Schema schema;
+        private T deserialized;
+
+        protected LazyImmutableContainer(byte[] serialized, Schema schema) {
+            this.serialized = serialized;
+            this.schema = schema;
+        }
+
+        protected T deserialized() {
+            if (deserialized == null) {
+                deserialized = deserialize(new GenericDatumReader<>(schema), serialized);
+            }
+            return deserialized;
+        }
+
+        @Override
+        public Schema getSchema() {
+            return schema;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (this == obj) {
+                return true;
+            }
+            if (!(obj instanceof GenericContainer
+                    && ((GenericContainer) obj).getSchema().getType() == schema.getType())) {
+                return false;
+            }
+            return deserialized().equals(obj);
+        }
+
+        @Override
+        public int hashCode() {
+            return deserialized().hashCode();
+        }
+    }
+
+    //region Lazy Immutable Containers
+    private static final class LazyImmutableRecord extends LazyImmutableContainer<GenericRecord>
+            implements GenericRecord {
+
+        private LazyImmutableRecord(byte[] serializedRecord, Schema schema) {
+            super(serializedRecord, schema);
+        }
+
+        @Override
+        public Object get(int i) {
+            return deserialized().get(i);
+        }
+
+        @Override
+        public Object get(String key) {
+            return deserialized().get(key);
+        }
+
+        @Override
+        public void put(String key, Object v) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void put(int i, Object v) {
+            throw new UnsupportedOperationException();
```

Comment:
will this not break Jet jobs using  generated classes (`SpecificRecordBase`) which might expect that they are mutable?

Reply:
the same for Array

Reply:
a5c5733e